### PR TITLE
Integrate DeepGEMM fp8 grouped gemm

### PR DESCRIPTION
**Pull request by yizhuoliang**

### CODE REVIEW

File: disagmoe/models/experts.py [hunk 1]
```diff
@@ -48,6 +62,45 @@ def create_weights(self, params_dtype: torch.dtype):
...
+        # w13_weight: [E, H, I*2] -> [E, I*2, H] for deep_gemm (if NT, B is transposed)
+        # w2_weight: [E, I, H] -> [E, H, I]
+        
+        # Scaling factors logic: this doesn't affect performance, and weights are 
+        # usually pre-quantized, so we just use ones for now.
+        # ceil_div = lambda x, y: (x + y - 1) // y
+        def ceil_div(x, y): return (x + y - 1) // y
+        
+        # w13
+        self.w13_weight_fp8 = self.w13_weight.transpose(1, 2).contiguous().to(torch.float8_e4m3fn)
```
Comment (shawlleyw) [inline disagmoe/models/experts.py]: We should allocate w13 and w2 with column first major, instead of transposing it before every execution.
Meanwhile I believe torch matmul and vllm linear method support both layouts.

File: disagmoe/models/experts.py [hunk 2]
```diff
@@ -48,6 +62,45 @@ def create_weights(self, params_dtype: torch.dtype):
...
+        self.w13_sfb = torch.ones(
+            self.num_experts, 
+            ceil_div(n_w13, 128), 
+            ceil_div(k_w13, 128), 
+            device=self.w13_weight.device, 
+            dtype=torch.float32
+        )
+        
+        # w2
+        self.w2_weight_fp8 = self.w2_weight.transpose(1, 2).contiguous().to(torch.float8_e4m3fn)
```
Comment (shawlleyw) [inline disagmoe/models/experts.py]: The same as w13 above

**Comment by shawlleyw** [comment]
overall current implementation is inefficient, we can tune it up

**Comment by shawlleyw** [comment]
The masked version can be used in cuda graph. m_grouped_fp8_gemm_nt_masked

File: disagmoe/models/experts.py [hunk 3]
```diff
@@ -48,9 +131,50 @@ def create_weights(self, params_dtype: torch.dtype):
         
         self.act_fn = torch.nn.SiLU(inplace=True)
 
+    def prepare_deep_gemm_fp8_weights(self):
```
Comment (shawlleyw) [inline disagmoe/models/experts.py]: Instead of transposing the weights, we can create another class that initialize weights following deepgemm's pattern.

File: disagmoe/models/experts.py [hunk 4]
```diff
@@ -77,6 +204,137 @@ def forward(self, bs: int, hiddens: torch.Tensor, batch_sizes: torch.Tensor):
             down = self.gmm(up, self.w2_weight, batch_sizes)
             output = down
         return output
+    
+    def _forward_deep_gemm(self, bs: int, hiddens: torch.Tensor, batch_sizes: torch.Tensor):
+        # When NON_GRAPH_NON_MASK is enabled, use the legacy non-graph, non-masked
+        # DeepGEMM implementation using the contiguous grouped kernel.
+        if NON_GRAPH_NON_MASK:
```
Comment (shawlleyw) [inline disagmoe/models/experts.py]: Separate the code for mask and non-masked version

File: disagmoe/models/experts.py (hunk 3 above)
Comment (yizhuoliang) [reply disagmoe/models/experts.py]: done

File: disagmoe/models/experts.py (hunk 4 above)
Comment (yizhuoliang) [reply disagmoe/models/experts.py]: done
